lib/SILGen/SILGenExpr.cpp

```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
+      return true;
+    if (!ty->getOptionalObjectType())
+      return false;
+    return ty->getOptionalObjectType()->isAnyClassReferenceType() ||
+           ty->getOptionalObjectType()->isVoid();
+  };
+
+  // Objective-C property.
+  if (auto *memberRefExpr = dyn_cast<MemberRefExpr>(E)) {
+    if (!hasAcceptableOuputType(memberRefExpr))
+      return nullptr;
+    auto *decl = memberRefExpr->getMember().getDecl();
+    auto *var = dyn_cast<VarDecl>(decl);
+    if (!var)
+      return nullptr;
+    if (!var->hasClangNode())
+      return nullptr;
+    auto *self = memberRefExpr->getBase();
+    if (!self->getType()->isAnyClassReferenceType())
+      return nullptr;
+    if (auto *load = dyn_cast<LoadExpr>(self))
+      self = load->getSubExpr();
+    auto *bindOptional = dyn_cast<BindOptionalExpr>(self);
+    if (!bindOptional)
+      return nullptr;
+    return bindOptional;
+  }
+
+  // Objective-C method call.
+  if (auto *call = dyn_cast<CallExpr>(E)) {
+    if (!hasAcceptableOuputType(call))
+      return nullptr;
+    auto *dotCall = dyn_cast<DotSyntaxCallExpr>(call->getDirectCallee());
+    if (!dotCall)
+      return nullptr;
+
+    auto *self = dotCall->getBase();
+    if (!self->getType()->isAnyClassReferenceType())
+      return nullptr;
+    if (auto *load = dyn_cast<LoadExpr>(self))
+      self = load->getSubExpr();
+    self = dyn_cast<BindOptionalExpr>(self);
+    if (!self)
+      return nullptr;
+
+    auto *func = dyn_cast<DeclRefExpr>(dotCall->getFn());
+    if (!func)
+      return nullptr;
+    if (!func->getDecl()->isObjC())
+      return nullptr;
+
+    return cast<BindOptionalExpr>(self);
+  }
+
+  // Objective-C protocol call.
+  if (auto *openExistential = dyn_cast<OpenExistentialExpr>(E)) {
+    auto *existential = openExistential->getExistentialValue();
+    if (auto *load = dyn_cast<LoadExpr>(existential))
+      existential = load->getSubExpr();
+    existential = dyn_cast<BindOptionalExpr>(existential);
+    if (!existential)
+      return nullptr;
+    auto *call = dyn_cast<CallExpr>(openExistential->getSubExpr());
+    if (!call)
+      return nullptr;
+    if (!hasAcceptableOuputType(call))
+      return nullptr;
+    auto *dotCall = dyn_cast<DotSyntaxCallExpr>(call->getDirectCallee());
+    if (!dotCall)
+      return nullptr;
+
+    // Self must be the opened existential.
+    auto *self = dotCall->getBase();
+    if (self != openExistential->getOpaqueValue())
+      return nullptr;
+
+    auto *func = dyn_cast<DeclRefExpr>(dotCall->getFn());
+    if (!func)
+      return nullptr;
+    if (!func->getDecl()->isObjC())
+      return nullptr;
+    return cast<BindOptionalExpr>(existential);
+  }
+  return nullptr;
+}
+
 RValue RValueEmitter::visitMemberRefExpr(MemberRefExpr *E, SGFContext C) {
   assert(!E->getType()->is<LValueType>() &&
          "RValueEmitter shouldn't be called on lvalues");
 
+
   if (isa<TypeDecl>(E->getMember().getDecl())) {
+    MarkBindOptionalShortCut shortCut(SGF, E);
```

Comment:
This is for a member access that is a type, eg something like type(of: value).SomeNestedClass. This doesn’t apply here.